autogen_pattern_macros.py

```diff
@@ -0,0 +1,61 @@
+import argparse
+
+def write_make_patterns(fp, n, i):
+    if(i == 1):
+        fp.write("#define MAKE_PATTERN_1(name, pattern_name, pattern, ...) static constexpr auto ptn_ ## name ## _1 = PATTERN(pattern);\n")
+        write_make_patterns(fp, n, i+1)
+    elif(i <= n):
+        fp.write("#define MAKE_PATTERN_%s(name, pattern_name, pattern, ...) static constexpr auto ptn_ ## name ## _%s = PATTERN(pattern); CONCATENATE(MAKE_PATTERN_%s(name, __VA_ARGS__),)\n" % (i, i, i-1))
+        write_make_patterns(fp, n, i+1)
+
+def write_name_patterns(fp, n, i):
```

Comment:
This needs a non-MSVC version (see how it's done in the existing ones, the difference is one comma in the end IIRC)